mitmproxy/net/udp.py

```diff
@@ -0,0 +1,328 @@
+from __future__ import annotations
```

Comment:
While the primary goal was to keep changes to mitmproxy at a minimum (~ +3k/-100), it also felt that this approach - while I wholeheartedly agree that it _does_ require in-depth knowledge of asyncio - is actually quite clean, in the sense that it doesn't dilute mitmproxy's existing concept of a connection. I've actually first experimented with another concept that would replace the `Address` as connection identifier with a generic, layer-dependent 'classifier'. The problem, however, was that this does not bode well with transparent mode. E.g. with DNS, a client could use the same id with different target servers. While I'm sure that this too could have been worked out, changes to the code base would - in generell - not have been trivial.

QUICs connection migration btw. could actually be implemented similar to the way packet routing is implemented in `UdpServer.datagram_received` and `UdpServerConnection.datagram_received`. Once a new connection is made (I think it's okay if migration is also logged as a new connection) and gets to the QUIC layer, if the connection ID of the first packet matches an existing one, the flow of the old connection is migrated to the new one (replace `client_conn`, check for matching `server_conn`, same problem as mentioned before with DNS in transparent mode), and the old connection is closed.

Another advantage of keeping it to a `base_events.Server` abstraction is, that if we ever plan to support listening on multiple ports, we could actually build a simple aggregating `base_events.Server`, as it already supports multiple sockets.

Reply:
To provide more constructive and concrete feedback on how we might continue from this point:

If we stick with the asyncio mimicry solution, we could:
- Modify the `DNSLayer` to close connections once the reply has been sent (since post-Kaminsky ports shouldn't be reused anyways).
- Make the connection watchdog timeout faster on UDP (with some inspiration from rfc4787: 2min).

If we decide to redo `Connection`s, I'd suggest:
1. Introducing a `ServerView` class:
  For everything except transparent mode, there will be just one instance (unless we one day provide the ability to listen on multiple ports). In transparent mode, however, whenever a client connects to a new ip:port pair, a new `ServerView` for that server is created. Since the concept of listen ports applies to both UDP and TCP, this approach (i.e. using ip and port) should be fine.
2. Providing a new `LogicalConnection` class, that is based on a layer dependent identifier.
   For HTTP1 this would be the clients ip:port pair, QUIC and DNS have an id.
3. Associating `Connection` instances with the new `LogicalConnection` class. Maybe also specify some `allow_multiple_connections` flag, that, if not set, will ensure that old client `Connection`s get closed.

The tricky part will be designing `ServerView`, that all the existing next layer logic works together with associating `Connection` to `LogicalConnection`. Incoming data of new `Connection`s needs to go through `ServerView` first to decide the `LogicalConnection`, then be passed on to the layer of that `LogicalConnection`. Unless layers are provided with a `ServerView` in the context and assign the client `Connection` to a `LogicalConnection` on their own. (Which actually might be a better choice, as it fits with the current hook mechanism and allows to alter the decision where to assign the `Connection`.)